File: frontend/src/utils/validation.js
/**
 * Validation utilities for input normalization and sanitization.
 */

/**
 * Normalize text input:
 * - Trim leading/trailing whitespace
 * - Normalize unicode to NFC form
 * - Return null if empty after normalization
 *
 * @param {string|null|undefined} text - Input string to normalize
 * @returns {string|null} Normalized string or null if empty
 *
 * @example
 * normalizeText("  Hello  ") // "Hello"
 * normalizeText("   ") // null
 * normalizeText(null) // null
 */
export function normalizeText(text) {
  if (!text) return null;

  // Trim whitespace
  text = text.trim();

  // Return null if empty
  if (!text) return null;

  // Normalize unicode to NFC form (canonical composition)
  // This ensures consistent representation of accented characters
  text = text.normalize("NFC");

  return text;
}

/**
 * Normalize array of strings:
 * - Apply normalizeText to each item
 * - Remove empty values
 * - Remove case-insensitive duplicates (keeps first occurrence)
 *
 * @param {string[]|null|undefined} items - Array of strings to normalize
 * @returns {string[]} Normalized, deduplicated array
 *
 * @example
 * normalizeArrayField(["  Naruto  ", "naruto", "Sasuke"]) // ["Naruto", "Sasuke"]
 * normalizeArrayField(["", "  ", "Valid"]) // ["Valid"]
 * normalizeArrayField(null) // []
 */
export function normalizeArrayField(items) {
  if (!items || !Array.isArray(items)) return [];

  const normalized = [];
  const seenLower = new Set();

  for (const item of items) {
    const normalizedItem = normalizeText(item);
    if (normalizedItem) {
      const lowerItem = normalizedItem.toLowerCase();
      if (!seenLower.has(lowerItem)) {
        normalized.push(normalizedItem);
        seenLower.add(lowerItem);
      }
    }
  }

  return normalized;
}

File: backend/app/utils/validation.py
"""
Validation utilities for input normalization and sanitization.
"""

import unicodedata
from typing import List, Optional


def normalize_text(text: Optional[str]) -> Optional[str]:
    """
    Normalize text input:
    - Strip leading/trailing whitespace
    - Normalize unicode to NFC form (canonical composition)
    - Return None if empty after normalization

    Args:
        text: Input string to normalize

    Returns:
        Normalized string or None if empty

    Examples:
        >>> normalize_text("  Hello  ")
        "Hello"
        >>> normalize_text("   ")
        None
        >>> normalize_text(None)
        None
    """
    if not text:
        return None

    # Strip whitespace
    text = text.strip()

    # Return None if empty after stripping
    if not text:
        return None

    # Normalize unicode to NFC form (canonical composition)
    # This ensures consistent representation of accented characters
    text = unicodedata.normalize("NFC", text)

    return text


def normalize_array_field(items: Optional[List[str]]) -> List[str]:
    """
    Normalize array of strings:
    - Apply normalize_text to each item
    - Remove None/empty values
    - Remove case-insensitive duplicates
    - Preserve original case of first occurrence

    Args:
        items: List of strings to normalize

    Returns:
        List of normalized, deduplicated strings

    Examples:
        >>> normalize_array_field(["  Naruto  ", "naruto", "Sasuke"])
        ["Naruto", "Sasuke"]
        >>> normalize_array_field(["", "  ", "Valid"])
        ["Valid"]
        >>> normalize_array_field(None)
        []
    """
    if not items:
        return []

    normalized = []
    seen_lower = set()

    for item in items:
        normalized_item = normalize_text(item)
        if normalized_item:
            # Check case-insensitive duplicates
            lower_item = normalized_item.lower()
            if lower_item not in seen_lower:
                normalized.append(normalized_item)
                seen_lower.add(lower_item)

    return normalized
